lib/db/backend/badger_backend.go

```diff
@@ -0,0 +1,377 @@
+	}
+	return val, nil
+}
+
+func (b *badgerBackend) NewPrefixIterator(prefix []byte) (Iterator, error) {
+	select {
+	case <-b.closed:
+		return nil, errClosed{}
+	default:
+	}
+
+	txn := b.bdb.NewTransaction(false)
+	it := badgerPrefixIterator(txn, prefix)
+	it.releaseFn = func() {
+		txn.Discard()
+	}
+	return it, nil
+}
+
+func (b *badgerBackend) NewRangeIterator(first, last []byte) (Iterator, error) {
+	select {
+	case <-b.closed:
+		return nil, errClosed{}
+	default:
+	}
+
+	txn := b.bdb.NewTransaction(false)
+	it := badgerRangeIterator(txn, first, last)
+	it.releaseFn = func() {
+		txn.Discard()
+	}
+	return it, nil
+}
+
+func (b *badgerBackend) Put(key, val []byte) error {
+	select {
+	case <-b.closed:
+		return errClosed{}
+	default:
+	}
+
+	txn := b.bdb.NewTransaction(true)
+	if err := txn.Set(key, val); err != nil {
+		txn.Discard()
+		return wrapBadgerErr(err)
+	}
+	return wrapBadgerErr(txn.Commit())
+}
+
+func (b *badgerBackend) Delete(key []byte) error {
+	select {
+	case <-b.closed:
+		return errClosed{}
+	default:
+	}
+
+	txn := b.bdb.NewTransaction(true)
+	if err := txn.Delete(key); err != nil {
+		txn.Discard()
+		return wrapBadgerErr(err)
+	}
+	return wrapBadgerErr(txn.Commit())
+}
+
+func (b *badgerBackend) Compact() error {
+	select {
+	case <-b.closed:
+		return errClosed{}
+	default:
+	}
+
+	// XXX: check if this is appropriate or if we also need Flatten and
+	// whatnot. Also it apparently returns an error if the GC didn't result
+	// in anything being freed...
+	_ = b.bdb.RunValueLogGC(0.5)
+	return nil
+}
+
+// badgerSnapshot implements backend.ReadTransaction
+type badgerSnapshot struct {
+	txn *badger.Txn
+	rel *releaser
+}
+
+func (l badgerSnapshot) Get(key []byte) ([]byte, error) {
+	item, err := l.txn.Get(key)
+	if err != nil {
+		return nil, wrapBadgerErr(err)
+	}
+	val, err := item.ValueCopy(nil)
+	if err != nil {
+		return nil, wrapBadgerErr(err)
+	}
+	return val, nil
+}
+
+func (l badgerSnapshot) NewPrefixIterator(prefix []byte) (Iterator, error) {
+	return badgerPrefixIterator(l.txn, prefix), nil
+}
+
+func (l badgerSnapshot) NewRangeIterator(first, last []byte) (Iterator, error) {
+	return badgerRangeIterator(l.txn, first, last), nil
+}
+
+func (l badgerSnapshot) Release() {
+	defer l.rel.Release()
+	l.txn.Discard()
+}
+
+// badgerTransaction implements backend.WriteTransaction using a batch (not
+// an actual badger transaction)
+type badgerTransaction struct {
+	badgerSnapshot
+	txn  *badger.Txn
+	bdb  *badger.DB
+	rel  *releaser
+	size int
+}
+
+func (t *badgerTransaction) Delete(key []byte) error {
+	t.size += len(key)
+	kc := make([]byte, len(key))
+	copy(kc, key)
+	return t.transactionRetried(func(txn *badger.Txn) error {
+		return txn.Delete(kc)
+	})
+}
+
+func (t *badgerTransaction) Put(key, val []byte) error {
+	t.size += len(key) + len(val)
+	kc := make([]byte, len(key))
+	copy(kc, key)
+	vc := make([]byte, len(val))
+	copy(vc, val)
+	return t.transactionRetried(func(txn *badger.Txn) error {
```

Comment:
Why do we need these copies?

Reply:
txn.Set is documented to keep references to the given slices and forbids modification of them until the transaction is committed, which is not the interface we promise in Backend I thing.

Reply:
I don't think we say anything about the values in either `Set` nor `Get` in the interface comments, but I also do think we treat it that way in code. So it should be added to the "backend specs" that after returning the given/returned values should be safe for modifying.

Reply:
The one thing that comes to mind is that we in many places go to some length to reuse the same key buffer repeatedly in a loop. For values I don't think we do the same, but still the interface for now is clearly that a database operation doesn't take ownership of the passed slices like the Badger operations do.

Then again that key operation is now meaningless with this interface so we could simplify our code elsewhere by always allocating and achieving the same thing... But I don't want to make this change bigger than necessary and Nobody(tm) will be running this code on day one.